File: js/packages/teams-ai/src/authentication/TeamsSsoBotAuthentication.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ActivityTypes, Storage, TokenResponse, TurnContext, tokenExchangeOperationName } from 'botbuilder';
import {
    Dialog,
    DialogContext,
    DialogTurnStatus,
    DialogSet,
    DialogState,
    DialogTurnResult,
    WaterfallDialog
} from 'botbuilder-dialogs';
import { Application } from '../Application';
import { TurnState } from '../TurnState';
import { BotAuthenticationBase } from './BotAuthenticationBase';
import { TeamsSsoPrompt } from './TeamsBotSsoPrompt';
import { TurnStateProperty } from '../TurnStateProperty';
import { ConfidentialClientApplication } from '@azure/msal-node';
import { TeamsSsoSettings } from './TeamsSsoSettings';

const SSO_DIALOG_ID = '_TeamsSsoDialog';

/**
 * @internal
 *
 * Handles authentication for Teams bots using Single Sign-On (SSO).
 * @template TState - The type of the turn state object.
 */
export class TeamsSsoBotAuthentication<TState extends TurnState> extends BotAuthenticationBase<TState> {
    private _prompt: TeamsSsoPrompt;
    private _tokenExchangeIdRegex: RegExp;

    /**
     * Initializes a new instance of the TeamsSsoBotAuthentication class.
     * @param {Application<TState>} app - The application object.
     * @param {TeamsSsoSettings} settings - The settings for Teams SSO.
     * @param {string} settingName - The name of the setting.
     * @param {ConfidentialClientApplication} msal - The MSAL (Microsoft Authentication Library) object.
     * @param {Storage} storage - The storage object for storing state.
     */
    public constructor(
        app: Application<TState>,
        settings: TeamsSsoSettings,
        settingName: string,
        msal: ConfidentialClientApplication,
        storage?: Storage
    ) {
        super(app, settingName, storage);

        this._prompt = new TeamsSsoPrompt('TeamsSsoPrompt', settingName, settings, msal);
        this._tokenExchangeIdRegex = new RegExp(
            `[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}-${this._settingName}`
        );

        // Do not save state for duplicate token exchange events to avoid eTag conflicts
        app.turn('afterTurn', async (context, state) => {
            return state.temp.duplicateTokenExchange !== true;
        });
    }

    /**
     * Run or continue the SSO dialog and returns the result.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} A promise that resolves to the dialog turn result containing the token response.
     */
    public async runDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>> {
        const dialogContext = await this.createSsoDialogContext(context, state, dialogStateProperty);
        let results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            results = await dialogContext.beginDialog(SSO_DIALOG_ID);
        }
        return results;
    }

    /**
     * Continues the SSO dialog and returns the result.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} A promise that resolves to the dialog turn result containing the token response.
     */
    public async continueDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>> {
        const dialogContext = await this.createSsoDialogContext(context, state, dialogStateProperty);
        return await dialogContext.continueDialog();
    }

    /**
     * Determines whether the token exchange activity should be processed by current authentication setting.
     * @param {TurnContext} context - The turn context object.
     * @returns {Promise<boolean>} A promise that resolves to a boolean indicating whether the token exchange route should be processed by current class instance.
     */
    public async tokenExchangeRouteSelector(context: TurnContext): Promise<boolean> {
        return (
            (await super.tokenExchangeRouteSelector(context)) &&
            this._tokenExchangeIdRegex.test(context.activity.value.id)
        );
    }

    /**
     * Creates the SSO dialog context.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogContext>} A promise that resolves to the dialog context object.
     */
    private async createSsoDialogContext(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogContext> {
        const accessor = new TurnStateProperty<DialogState>(state, 'conversation', dialogStateProperty);
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this._prompt);
        dialogSet.add(
            new WaterfallDialog(SSO_DIALOG_ID, [
                async (step) => {
                    return await step.beginDialog(this._prompt.id);
                },
                async (step) => {
                    const tokenResponse = step.result;
                    // Dedup token exchange responses
                    if (tokenResponse && (await this.shouldDedup(context))) {
                        state.temp.duplicateTokenExchange = true;
                        return Dialog.EndOfTurn;
                    }
                    return await step.endDialog(step.result);
                }
            ])
        );
        return await dialogSet.createContext(context);
    }

    /**
     * Checks if deduplication should be performed for token exchange.
     * @param {TurnContext} context - The turn context object.
     * @returns {Promise<boolean>} A promise that resolves to a boolean indicating whether deduplication should be performed.
     */
    private async shouldDedup(context: TurnContext): Promise<boolean> {
        const storeItem = {
            eTag: context.activity.value.id
        };

        const key = this.getStorageKey(context);
        const storeItems = { [key]: storeItem };

        try {
            await this._storage.write(storeItems);
        } catch (err) {
            if (err instanceof Error && err.message.indexOf('eTag conflict')) {
                return true;
            }
            throw err;
        }
        return false;
    }

    /**
     * Gets the storage key for storing the token exchange state.
     * @param {TurnContext} context - The turn context object.
     * @returns {string} The storage key.
     * @throws Error if the context is invalid or the activity is not a token exchange invoke.
     */
    private getStorageKey(context: TurnContext): string {
        if (!context || !context.activity || !context.activity.conversation) {
            throw new Error('Invalid context, can not get storage key!');
        }
        const activity = context.activity;
        const channelId = activity.channelId;
        const conversationId = activity.conversation.id;
        if (activity.type !== ActivityTypes.Invoke || activity.name !== tokenExchangeOperationName) {
            throw new Error('TokenExchangeState can only be used with Invokes of signin/tokenExchange.');
        }
        const value = activity.value;
        if (!value || !value.id) {
            throw new Error('Invalid signin/tokenExchange. Missing activity.value.id.');
        }
        return `${channelId}/${conversationId}/${value.id}`;
    }
}

File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/Authentication/Bot/TeamsSsoBotAuthentication.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Teams.AI.State;
using Microsoft.Teams.AI.Exceptions;
using Newtonsoft.Json.Linq;
using System.Text.RegularExpressions;

namespace Microsoft.Teams.AI
{
    /// <summary>
    /// Handles authentication for bot in Teams using Teams SSO.
    /// </summary>
    internal class TeamsSsoBotAuthentication<TState> : BotAuthenticationBase<TState>
        where TState : TurnState, new()
    {
        private const string SSO_DIALOG_ID = "_TeamsSsoDialog";
        private Regex _tokenExchangeIdRegex;
        protected TeamsSsoPrompt _prompt;

        /// <summary>
        /// Initializes the class
        /// </summary>
        /// <param name="app">The application instance</param>
        /// <param name="name">The name of current authentication handler</param>
        /// <param name="settings">The authentication settings</param>
        /// <param name="storage">The storage to save turn state</param>
        public TeamsSsoBotAuthentication(Application<TState> app, string name, TeamsSsoSettings settings, IStorage? storage = null) : base(app, name, storage)
        {
            _tokenExchangeIdRegex = new Regex("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}-" + name);
            _prompt = new TeamsSsoPrompt("TeamsSsoPrompt", name, settings);

            // Do not save state for duplicate token exchange events to avoid eTag conflicts
            app.OnAfterTurn((context, state, cancellationToken) =>
            {
                return Task.FromResult(state.Temp.DuplicateTokenExchange != true);
            });
        }

        /// <summary>
        /// Continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public override async Task<DialogTurnResult> ContinueDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default)
        {
            DialogContext dialogContext = await CreateSsoDialogContext(context, state, dialogStateProperty);
            return await dialogContext.ContinueDialogAsync();
        }

        /// <summary>
        /// Run or continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public override async Task<DialogTurnResult> RunDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default)
        {
            DialogContext dialogContext = await CreateSsoDialogContext(context, state, dialogStateProperty);
            DialogTurnResult result = await dialogContext.ContinueDialogAsync();
            if (result.Status == DialogTurnStatus.Empty)
            {
                result = await dialogContext.BeginDialogAsync(SSO_DIALOG_ID);
            }
            return result;
        }

        /// <summary>
        /// The route selector for signin/tokenExchange activity
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>True if the activity should be handled by current authentication hanlder. Otherwise, false.</returns>
        internal override async Task<bool> TokenExchangeRouteSelector(ITurnContext context, CancellationToken cancellationToken)
        {
            JObject? value = context.Activity.Value as JObject;
            JToken? id = value?["id"];
            string idStr = id?.ToString() ?? "";
            return await base.TokenExchangeRouteSelector(context, cancellationToken)
                && this._tokenExchangeIdRegex.IsMatch(idStr);
        }

        private async Task<DialogContext> CreateSsoDialogContext(ITurnContext context, TState state, string dialogStateProperty)
        {
            TurnStateProperty<DialogState> accessor = new(state, "conversation", dialogStateProperty);
            DialogSet dialogSet = new(accessor);
            WaterfallDialog ssoDialog = new(SSO_DIALOG_ID);

            dialogSet.Add(this._prompt);
            dialogSet.Add(new WaterfallDialog(SSO_DIALOG_ID, new WaterfallStep[]
            {
                async (step, cancellationToken) =>
                {
                    return await step.BeginDialogAsync(this._prompt.Id);
                },
                async (step, cancellationToken) =>
                {
                    TokenResponse? tokenResponse = step.Result as TokenResponse;
                    if (tokenResponse != null && await ShouldDedup(context))
                    {
                        state.Temp.DuplicateTokenExchange = true;
                        return Dialog.EndOfTurn;
                    }
                    return await step.EndDialogAsync(step.Result);
                }
            }));
            return await dialogSet.CreateContextAsync(context);
        }

        private async Task<bool> ShouldDedup(ITurnContext context)
        {
            string key = GetStorageKey(context);
            string id = (context.Activity.Value as JObject)?.Value<string>("id")!; // The id exists if GetStorageKey success
            IStoreItem storeItem = new TokenStoreItem(id);
            Dictionary<string, object> storesItems = new()
            {
                {key, storeItem}
            };

            try
            {
                await this._storage.WriteAsync(storesItems);
            }
            catch (Exception ex)
            {
                if (ex.Message.StartsWith("Etag conflict", StringComparison.OrdinalIgnoreCase) || ex.Message.Contains("pre-condition is not met"))
                {
                    return true;
                }
                throw;
            }
            return false;
        }

        private string GetStorageKey(ITurnContext context)
        {
            if (context == null || context.Activity == null
                || context.Activity.Conversation == null)
            {
                throw new AuthException("Invalid context, can not get storage key!");
            }

            Activity activity = context.Activity;
            string channelId = activity.ChannelId;
            string conversationId = activity.Conversation.Id;
            if (activity.Type != ActivityTypes.Invoke || activity.Name != SignInConstants.TokenExchangeOperationName)
            {
                throw new AuthException("TokenExchangeState can only be used with Invokes of signin/tokenExchange.");
            }
            JObject value = JObject.FromObject(activity.Value);
            JToken? id = value["id"];
            if (id == null)
            {
                throw new AuthException("Invalid signin/tokenExchange. Missing activity.value.id.");
            }
            return $"{channelId}/{conversationId}/{id}";
        }
    }

    internal class TokenStoreItem : IStoreItem
    {
        public string ETag { get; set; }

        public TokenStoreItem(string etag)
        {
            ETag = etag;
        }
    }
}
